### PR TITLE
[MM-41109] Make all links from plugin settings open in a different tab

### PR DESCRIPTION
**Pull request by Borknab**
#### Summary

This PR changes the admin_console's `SchemaText` rendering logic, so that links on plugins' settings pages open in external tabs.  Consequently, `FormattedMarkdownMessage` component also needed to change in order to support making all the links within markdown text to open in external tabs.

#### Ticket Link

Fixes https://github.com/mattermost/mattermost-server/issues/19405
https://mattermost.atlassian.net/browse/MM-41109

### CODE REVIEW

**Review by hmhealey — COMMENTED** [review]
Thanks for the PR! One concern I have with this is that, since SchemaText is used to render all help text in the System Console, it makes every link in it always open in a new tab. I don't think we want that to apply to all links since System Console pages often link to each other which we'd want to do in the current tab.

What if, instead of making all links open in a new tab, we instead use [the same logic that's used for Markdown in posts](https://github.com/mattermost/mattermost-webapp/blob/master/utils/markdown/renderer.tsx#L158) for `FormattedMarkdownMessage` so that external links open in a new tab and internal ones open in the current tab? That could actually make it so that we don't need to keep prefixing those URLs with an exclamation mark to do that either since I can't remember exactly why we chose to do it that way in the first place.

**Comment by Borknab** [comment]
@hmhealey good suggestion, thanks!
I implemented it, so now all things using `FormattedMarkdownMessage` should use this logic: `shouldOpenInNewTab()` decides whether the links opens in new tab or not. The feature with `[](!foo)` for external links should stay I think, as it is probably used in many files.

File: components/admin_console/schema_text.jsx [hunk 1]
```diff
@@ -38,6 +38,8 @@ export default class SchemaText extends React.PureComponent {
 
         if (isMarkdown) {
             return (
+
+                // Give all markdown links target="_blank"
```
Comment (mickmister) [inline components/admin_console/schema_text.jsx:42]: This comment seems a little unrelated to the purpose of this file/block of code, and adds a little bit of noise to readability. Thoughts on removing this comment @Borknab?

File: components/formatted_markdown_message.jsx [hunk 2]
```diff
@@ -17,12 +19,18 @@ export class CustomRenderer extends marked.Renderer {
...
+
         if (this.disableLinks) {
             return text;
         }
         if (href[0] === TARGET_BLANK_URL_PREFIX) {
             return `<a href="${href.substring(1, href.length)}" rel="noopener noreferrer" target="_blank">${text}</a>`;
         }
+        if (openInNewTab) {
+            return `<a href="${href}" target="_blank">${text}</a>`;
+        }
```
Comment (mickmister) [inline components/formatted_markdown_message.jsx:33]: I'm thinking we should add the same `rel="noopener noreferrer"` attribute for this new addition.

**Review by mickmister — CHANGES_REQUESTED** [review]
Awesome work @Borknab! I have two minor suggestions on the changes. Otherwise LGTM!

File: components/admin_console/schema_text.jsx (hunk 1 above)
Comment (Borknab) [reply components/admin_console/schema_text.jsx]: Agreed. Doesn't make sense, thanks for pointing that out!

File: components/formatted_markdown_message.jsx (hunk 2 above)
Comment (Borknab) [reply components/formatted_markdown_message.jsx]: After thinking about this, it makes sense, thanks!

**Comment by Borknab** [comment]
@mickmister I adjusted the code according to your suggestions, I totally agree with them. Thanks!

**Review by mickmister — APPROVED** [review]
LGTM :tada: Thanks @Borknab!

**Review by hmhealey — APPROVED** [review]
Sorry for the delay in review. This looks good to me.

It'll also be nice to remove the exclamation marks from most of the URLs throughout the System Console help text as well since that won't be necessary any more. We've got a parallel effort to remove most of those links from the translation strings though, so we shouldn't make further changes to that area right now.

**Comment by mickmister** [comment]
@DHaussermann Maybe assign this to @dipak-demansol for testing?

**Comment by dipak-demansol** [comment]
> @DHaussermann Maybe assign this to @dipak-demansol for testing?

@mickmister Sure, You can assign me this.

**Comment by mickmister** [comment]
@dipak-demansol It won't let me request your review for some reason so I assigned you instead

**Comment by dipak-demansol** [comment]
@mickmister LGTM,
:- Tested on Browser and desktop app, Tested when The browser was already open and when it was not open.

non-blocking point
---------------------------
:- when the browser is not opened and i click on the link from the desktop app then the browser is opened but the browsers "X", minimize, maximize" button is not visible when browser is full screen, same issue not generated when browser was already open. (i know we don't have Control on Browser but still for your info, i reported what i found)

**Comment by dipak-demansol** [comment]
> > > :- when the browser is not opened and i click on the link from the desktop app then the browser is opened but the browsers "X", minimize, maximize" button is not visible when browser is full screen, same issue not generated when browser was already open. (i know we don't have Control on Browser but still for your info, i reported what i found)
> > 
> > 
> > @dipak-demansol Does this behavior differ from what happens when you click a similar link that's in a post, when using the desktop app?
> 
>  vokoscreenNG-2022-04-26_19-37-25.mov 
> No, it has the same behavior, here is the video of that.

@mickmister :arrow_up:

**Comment by mickmister** [comment]
@dipak-demansol If clicking a link in a post has the same behavior as clicking a link in the system console, then this can be approved.

**Review by DHaussermann — APPROVED** [review]
Tested and approved by @dipak-demansol

**Comment by hmhealey** [comment]
Thanks again for the PR, @Borknab!

**Comment by Borknab** [comment]
@hmhealey no problem, was glad to contribute! :)

**Pull request merged**